### DISCUSSION TITLE
GitHub Actions stuck waiting for a self hosted runner on Container App Jobs

### DISCUSSION THREAD

**Discussion by krusty93** [community/community#84635] [Actions]
### Select Topic Area

Bug

### Body

We actually use self hosted runners to run some of our GitHub Actions. Runners are hosted on Azure Container Apps Jobs as described in the [Microsoft documentation](https://learn.microsoft.com/en-us/azure/container-apps/tutorial-ci-cd-runners-jobs?tabs=bash&pivots=container-apps-jobs-self-hosted-ci-cd-github-actions).

However, sometimes we hit a strange behavior of the runners.

When a GitHub Actions starts using the self hosted runner, it basically works. However, at the end of the execution the runner won't stop but instead keeps running (I can see the job running from Azure Portal).

When a new GitHub Actions is triggered (or the same Action starts a new job after the first one) requiring a new runner instance, most of the times the runner won't register itself in the repo and the pipeline stuck there forever.
The only way to get out of this is to use the CLI command `az containerapp job stop --job-execution-name <container-name> -g <rg> -n <container-app>` forcing the runner to stop. Then, a new instance starts and is ready to pick up the waiting Actions and get the job done.

Finally, we never see the runner shutting down gracefully: the Azure Portal always reports "Failed" as per job status, even if the Actions completes successfully.

[Here you can find](https://github.com/pagopa/github-self-hosted-runner-azure/blob/e3a5b35f0545426b19144fe385ca03dd326f67bd/Dockerfile.github#L1C37-L1C44) the Docker image we've been using (extends the v2.311.0 adding some binaries).

The `github-runner-entrypoint.sh` used is available [here](https://github.com/pagopa/github-self-hosted-runner-azure/blob/e3a5b35f0545426b19144fe385ca03dd326f67bd/github-runner-entrypoint.sh#L1).

[I attach the logs ](https://github.com/community/community/files/13844656/query_data.csv) extracted from the Container App job using this query:
```
ContainerAppConsoleLogs_CL 
| where ContainerName_s == '...'
| where Log_s has 'INFO Runner'
| project TimeGenerated, ContainerGroupName_s, ContainerId_s, Log_s
| order by TimeGenerated desc
```
They look like an exception stack trace but I can't figure out where and why it's coming from since it's runner internal code.

Any help would be appreciated

**Comment by KNCn23** [comment]
It seems you're encountering a persistent issue with GitHub Actions self-hosted runners hosted on Azure Container Apps. The primary problem is that the runners don't stop at the end of their job execution and don't shut down gracefully, leading to a stuck pipeline and the need to force-stop the runner using Azure CLI.

To address this issue, a few areas should be investigated:

Runner Configuration: Ensure that your self-hosted runners are configured correctly according to the GitHub Actions and Azure Container Apps documentation. This includes verifying environment variables, permissions, and network configurations.
Docker Image and Entrypoint Script: Since you're using a custom Docker image (extending v2.311.0), ensure that the modifications or additional binaries included don't interfere with the normal operation of the runner. Similarly, review the github-runner-entrypoint.sh script for any potential issues that might prevent the runner from shutting down properly.
Logs Analysis: The logs you mentioned, which look like an exception stack trace, are crucial. Analyze these logs to identify any recurring error patterns or exceptions that could indicate why the runner isn't stopping as expected.
GitHub and Azure Updates: Ensure that both your GitHub Actions runner version and Azure Container Apps are up to date. Sometimes, such issues are resolved in newer releases.
Community and Support Channels: Check GitHub’s community forums and Azure’s support channels. Other users might have faced similar issues and could have shared solutions or workarounds.
Testing with a Basic Configuration: Try running a very basic runner setup without any customizations to see if the issue persists. This can help isolate whether the problem is with your specific setup or a more general issue with the runner or Azure Container Apps.
Resource Allocation and Limits: Inspect if there are any resource limits on the Azure side that might be causing the runners to behave unexpectedly, especially if the issue occurs after long-running jobs or heavy resource usage.
Alternative Approaches: If the issue persists, consider alternative hosting options for your runners, or using a different CI/CD setup that might be more stable.
Contacting Support: If none of the above steps help, consider reaching out to GitHub or Azure support for more in-depth assistance.
Given the complexity of the issue, it might require a combination of these approaches to diagnose and resolve. Remember to document any changes you make during troubleshooting, as this can be helpful for future reference or if you need to escalate the issue to support.

**Comment by KNCn23** [comment]
If this article helped you, can you mark both of my messages as answers?
